### PR TITLE
✨ Add weighted avg

### PR DESCRIPTION
**Pull request by Rubilmax**

### CODE REVIEW

File: src/math/PercentageMath.sol [hunk 1]
```diff
@@ -51,4 +56,19 @@ library PercentageMath {
             y := div(add(mul(x, PERCENTAGE_FACTOR), y), percentage)
         }
     }
+
+    /// @notice Executes a percent average, given an interval [x, y] and a percent p: x * (1 - p) + y * p
+    /// @param x The value the start of the interval (included).
+    /// @param y The value the end of the interval (included).
```
Comment (MerlinEgalite) [inline src/math/PercentageMath.sol:62]: do re-write I guess

File: test/TestPercentageMath.sol [hunk 2]
```diff
@@ -24,6 +32,16 @@ contract PercentageMathFunctionsRef {
...
+
+    function percentAvg(
+        uint256 x,
+        uint256 y,
+        uint256 percentage
+    ) public pure returns (uint256) {
+        return
+            PercentageMathRef.percentMul(x, PercentageMathRef.PERCENTAGE_FACTOR - percentage) +
+            PercentageMathRef.percentMul(y, percentage);
+    }
```
Comment (MerlinEgalite) [inline test/TestPercentageMath.sol:44]: Use the same no? As it will be the reference after

File: test/TestPercentageMath.sol [hunk 3]
```diff
@@ -64,6 +82,37 @@ contract TestPercentageMath is Test {
         assertEq(PercentageMath.percentDiv(x, y), PercentageMathRef.percentDiv(x, y));
     }
 
+    function testPercentAvg(
+        uint256 x,
+        uint256 y,
+        uint16 percentage
+    ) public {
+        vm.assume(percentage <= PERCENTAGE_FACTOR);
+
```
Comment (MerlinEgalite) [inline test/TestPercentageMath.sol:91]: why the line space?

**Review by MerlinEgalite — CHANGES_REQUESTED** [review]
Some changes to add

File: src/math/PercentageMath.sol [hunk 4]
```diff
@@ -51,4 +56,19 @@ library PercentageMath {
...
+    /// @param percentage The percentage of the interval to be calculated.
+    /// @return the average of x and y, weighted by percentage.
+    function percentAvg(
+        uint256 x,
+        uint256 y,
+        uint256 percentage
+    ) internal pure returns (uint256) {
+        if (percentage > PercentageMath.PERCENTAGE_FACTOR) revert PercentageTooHigh();
+
+        return percentMul(x, PercentageMath.PERCENTAGE_FACTOR - percentage) + percentMul(y, percentage);
```
Comment (QGarchery) [inline src/math/PercentageMath.sol:72]:
```suggestion
        if (percentage > PERCENTAGE_FACTOR) revert PercentageTooHigh();

        return percentMul(x, PERCENTAGE_FACTOR - percentage) + percentMul(y, percentage);
```

File: test/TestPercentageMath.sol [hunk 5]
```diff
@@ -64,6 +82,37 @@ contract TestPercentageMath is Test {
...
+        vm.assume(
+            percentage < PercentageMathRef.PERCENTAGE_FACTOR &&
+                x <= MAX_UINT256_MINUS_HALF_PERCENTAGE / (PercentageMathRef.PERCENTAGE_FACTOR - percentage)
+        );
+        vm.assume(percentage > 0 && y <= MAX_UINT256_MINUS_HALF_PERCENTAGE / percentage);
+
+        assertEq(
+            PercentageMath.percentAvg(x, y, percentage),
+            PercentageMathRef.percentMul(x, PercentageMathRef.PERCENTAGE_FACTOR - percentage) +
+                PercentageMathRef.percentMul(y, percentage)
```
Comment (QGarchery) [inline test/TestPercentageMath.sol:104]: Why not using the reference implementation here ?

File: src/math/PercentageMath.sol [hunk 6]
```diff
@@ -51,4 +56,19 @@ library PercentageMath {
             y := div(add(mul(x, PERCENTAGE_FACTOR), y), percentage)
         }
     }
+
+    /// @notice Executes a percent average, given an interval [x, y] and a percent p: x * (1 - p) + y * p
+    /// @param x The value the start of the interval (included).
+    /// @param y The value the end of the interval (included).
+    /// @param percentage The percentage of the interval to be calculated.
+    /// @return the average of x and y, weighted by percentage.
+    function percentAvg(
```
Comment (MathisGD) [inline src/math/PercentageMath.sol:65]: weird naming for this IMO. Weighted average is more accurate no ?

File: test/TestPercentageMath.sol (hunk 2 above)
Comment (Rubilmax) [reply test/TestPercentageMath.sol:44]: This is what's done here, except I did not copy the revert

File: test/TestPercentageMath.sol [hunk 7]
```diff
@@ -64,6 +82,37 @@ contract TestPercentageMath is Test {
         assertEq(PercentageMath.percentDiv(x, y), PercentageMathRef.percentDiv(x, y));
     }
 
+    function testPercentAvg(
+        uint256 x,
+        uint256 y,
+        uint16 percentage
+    ) public {
+        vm.assume(percentage <= PERCENTAGE_FACTOR);
```
Comment (QGarchery) [inline test/TestPercentageMath.sol:94]: We can remove this line, no ?

Comment (Rubilmax) [reply test/TestPercentageMath.sol:94]: If this line gets removed then `percentage` is no longer bounded and can lead to a revert when passed to `percentAvg`

File: test/TestPercentageMath.sol (hunk 5 above)
Comment (Rubilmax) [reply test/TestPercentageMath.sol:104]: because there is no reference implementation in aave's contracts

File: test/TestPercentageMath.sol [hunk 8]
```diff
@@ -64,6 +86,36 @@ contract TestPercentageMath is Test {
...
+        vm.assume(
+            percentage < PercentageMathRef.PERCENTAGE_FACTOR &&
```
Comment (MathisGD) [inline test/TestPercentageMath.sol:96]:
```suggestion
            percentage < PERCENTAGE_FACTOR &&
```

File: src/math/PercentageMath.sol (hunk 4 above)
Comment (Rubilmax) [reply src/math/PercentageMath.sol:72]: oups

File: test/TestPercentageMath.sol [hunk 9]
```diff
@@ -64,6 +86,36 @@ contract TestPercentageMath is Test {
...
+                x <= MAX_UINT256_MINUS_HALF_PERCENTAGE / (PercentageMathRef.PERCENTAGE_FACTOR - percentage)
```
Comment (MathisGD) [inline test/TestPercentageMath.sol:97]:
```suggestion
                x <= MAX_UINT256_MINUS_HALF_PERCENTAGE / (PERCENTAGE_FACTOR - percentage)
```

**Review by MathisGD — CHANGES_REQUESTED** [review]

File: test/TestPercentageMath.sol (hunk 7 above)
Comment (QGarchery) [reply test/TestPercentageMath.sol:94]: Isn't it a duplicate of one of the checks of the line coming just after ?
```solidity
vm.assume(
            percentage < PERCENTAGE_FACTOR && x <= MAX_UINT256_MINUS_HALF_PERCENTAGE / (PERCENTAGE_FACTOR - percentage)
        );
```

File: test/TestPercentageMath.sol (hunk 5 above)
Comment (QGarchery) [reply test/TestPercentageMath.sol]: Let's say that the contract `PercentageMathFunctionsRef` you just wrote is the ref, is it possible to reuse that ?

**Review by QGarchery — APPROVED** [review]

File: test/TestPercentageMath.sol (hunk 7 above)
Comment (Rubilmax) [reply test/TestPercentageMath.sol:94]: oh you just highlighted a mistake: this line should be:
```solidity
vm.assume(
            percentage == PERCENTAGE_FACTOR || x <= MAX_UINT256_MINUS_HALF_PERCENTAGE / (PERCENTAGE_FACTOR - percentage)
        );
```

**Review by MerlinEgalite — APPROVED** [review]
Just a comment but LGTM

**Comment by QGarchery** [comment]
> Just a comment but LGTM

Where is this comment @MerlinEgalite ? I don't see it

**Comment by MerlinEgalite** [comment]
> > Just a comment but LGTM
> 
> Where is this comment @MerlinEgalite ? I don't see it

here https://github.com/morpho-labs/morpho-utils/pull/32#issuecomment-1202991579

**Comment by QGarchery** [comment]
> > > Just a comment but LGTM
> > 
> > 
> > Where is this comment @MerlinEgalite ? I don't see it
> 
> here [#32 (comment)](https://github.com/morpho-labs/morpho-utils/pull/32#issuecomment-1202991579)

This link points to my comment hehe

**Pull request merged**